File: Frontend/graph-ql/gql-id64.js
import { GqlType } from './gql-type.js'
import { nonNull } from './gql.js'

export const GqlId64 = new class GqlId64 extends GqlType {
  resolve (input) {
    if (input === null) { return null }
    if (typeof input !== 'string') { throw new TypeError(`Объект <${input}> должен иметь тип <GqlId64>`) }
    return input
  }

  serialize (object) {
    return object
  }

  validate (input) {
    if (input !== undefined && input !== null && typeof input !== 'string') { return 'Тип должен быть <GqlId64>' }
  }
}()

export const GqlNonNullId64 = nonNull(GqlId64)

File: Backend/Gql/Types/Scalars/GqlId64.cs
using System.Globalization;
using Newtonsoft.Json.Linq;

namespace Backend.Gql.Types.Scalars {
    public class GqlId64 : GqlScalar<Id64?> {
        public override object Parse(object value, Response context)
        {
            if (value is string token) {
                if (long.TryParse(token, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var number)) return new Id64(number);
                if (token == "null") return null;
            }
            context.AddError("", $"Could not convert input <{value}> string to number or null");
            return null;
        }

        public override object Resolve(Selection query, object obj, Response response)
        {
            return obj is Id64 id64 ? JToken.FromObject(id64) : null;
        }

        public override string BuildSchemeJavascript()
        {
            return "GqlId64";
        }
    }

    public class GqlNonNullId64 : GqlScalar<Id64> {
        public override object Parse(object value, Response context)
        {
            if (value is string token &&
                long.TryParse(token, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var number)) return new Id64(number);
            context.AddError("", $"Could not convert input <{value}> string to number");
            return null;
        }

        public override object Resolve(Selection query, object obj, Response response)
        {
            return obj is Id64 id64 ? JToken.FromObject(id64) : null;
        }

        public override string BuildSchemeJavascript()
        {
            return "GqlNonNullId64";
        }
    }
}